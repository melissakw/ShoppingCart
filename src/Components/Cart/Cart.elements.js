import styled from "styled-components";

export const Loading = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  min-height: 100%;
  background-color: black;
  color: black;
  opacity: 0.2;
  z-index: 100;
  text-align: center;
  padding-top: 50vh;
`;

export const Table = styled.table`

    border-spacing: 30px;

  @media only screen and (max-width: 600px) {
  table,
  thead,
  tbody,
  th,
  td,
  tr {
    text-align: left;
    display: block;
  }

  thead tr {
    text-align: left;
    position: absolute;
    top: -9999px;
    left: -9999px;
  }

  td {
    /* Makes td act like a "row" */
    border: none;
    position: relative;
    white-space: normal;
    text-align: left;
  }

  td:before {
    width: 45%;
    text-align: left;
    content: attr(data-title);

  }
}
`;

export const CartRow = styled.div`
  display: inline-flex;
  flex-wrap: wrap;
  width: 100%;
  justify-content: space-between;
  flex-direction: row;
`;

export const CartProductImage = styled.img`
  width: 150px;
`;

export const RemoveButton = styled.button`
  background-color: white;
  font-size: 14px;
  border: none;
  padding: 0;
  margin-top: 40px;
  overflow: hidden;
  cursor: pointer;
  opacity: 0.6;
  transition: 0.4s;

  &:hover {
    opacity: 1;
  }
`;

export const CartSummary = styled.div`
  margin-top: 60px;
  margin-right: 60px;
  height: 300px;
  padding: 30px;
  border: 2px lightgray solid;
`;

export const ShipTime = styled.p`
  font-size: 14px;
  font-style: italic;
`;

export const CheckoutButton = styled.button`
  background-color: black;
  color: white;
  width: 100%;
  font-size: 20px;
  border: none;
  padding: 10px;
  margin-top: 10px;
  overflow: hidden;
  cursor: pointer;
  opacity: 0.6;
  transition: 0.4s;

  &:hover {
    opacity: 1;
  }
`;
