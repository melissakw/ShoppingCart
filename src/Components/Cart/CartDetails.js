import React from "react";
import { useDispatch } from "react-redux";
import { removeItem, editItem } from "../../Redux/ActionCreators";
import {
  Table,
  CartProductImage,
  RemoveButton
} from "./Cart.elements";

const CartDetails = ({ productsInCart, setIsLoadingChanges }) => {
  const dispatch = useDispatch();

  return (
    <Table>
      <thead>
        <tr>
          <th></th>
          <th>Item</th>
          <th>Item Price</th>
          <th>Item Quantity</th>
        </tr>
      </thead>
      <tbody>
        {productsInCart.map((productInCart) => {
          const stockQty = Array(productInCart.stock)
            .fill()
            .map((_, i) => i + 1);

          return (
            <tr key={productInCart.id}>
              <td>
                <CartProductImage
                  src={productInCart.image}
                  alt={productInCart.title}
                />
              </td>
              <td>
                <p>{productInCart.title}</p>
              </td>
              <td>
                <p>${productInCart.price}</p>
              </td>
              <td>
                <div>
                  <select
                    name="quantity"
                    id="quantity"
                    defaultValue={productInCart.quantity}
                    onChange={(e) => {
                      setIsLoadingChanges(true);
                      dispatch(editItem(productInCart, e.target.value));
                      setTimeout(() => setIsLoadingChanges(false), 1800);
                    }}
                    style={{ marginTop: "58px" }}
                  >
                    {stockQty.map((num) => {
                      return <option key={num} value={num}>{num}</option>;
                    })}
                  </select>
                </div>
                <div>
                  <RemoveButton
                    onClick={() => {
                      setIsLoadingChanges(true);
                      dispatch(removeItem(productInCart.id));
                      setTimeout(() => setIsLoadingChanges(false), 1800);
                    }}
                  >
                    remove
                  </RemoveButton>
                </div>
              </td>
            </tr>
          );
        })}
      </tbody>
    </Table>
  );
};

export default CartDetails;
