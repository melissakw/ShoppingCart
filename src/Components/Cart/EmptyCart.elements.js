import styled from "styled-components";
import { Link } from "react-router-dom";

export const EmptyCartContainer = styled.div`
  padding: 20px;
  text-align: center;
`;

export const ShopLink = styled(Link)`
  color: black;
`;

export const HighlightText = styled.div`
  font-family: "Times New Roman", Times, serif;
  font-size: 34px;
  margin-bottom: 15px;
`;

export const BodyText = styled.div`
  font-size: 16px;
`;
