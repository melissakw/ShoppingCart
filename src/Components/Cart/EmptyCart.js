import React from "react";
import {
  EmptyCartContainer,
  ShopLink,
  HighlightText,
  BodyText
} from "./EmptyCart.elements";

const EmptyCart = () => {
  return (
    <EmptyCartContainer>
      <HighlightText>Your cart is empty</HighlightText>
      <ShopLink to="/">
        <BodyText>Find something you love to fill it up</BodyText>
      </ShopLink>
    </EmptyCartContainer>
  );
};

export default EmptyCart;
