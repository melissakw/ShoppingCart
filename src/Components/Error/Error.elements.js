import styled from "styled-components";

export const GoHomeButton = styled.button`
  padding: 10px;
  border: none;
  background-color: darkorange;
  color: white;
  font-weight: bold;
  transition: all 0.4s ease;
  opacity: 0.7;

  &:hover {
    letter-spacing: 1px;
    opacity: 1;
  }
`;

export const HighlightText = styled.h1`
  color: #838383;
  font-family: "Times New Roman", Times, serif;
`;

export const BodyText = styled.h3`
  font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
`;
