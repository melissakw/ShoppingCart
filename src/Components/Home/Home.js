import React, { useEffect, useState } from "react";
import ProductsPage from "../ProductsPage/ProductsPage";
import HeaderBar from "../HeaderBar/HeaderBar";
import axios from "axios";
import CircularProgress from "@material-ui/core/CircularProgress";
import { HomeContainer } from "./Home.elements";
import { AppContainer, LoadingContainer } from "../App.elements";

const Home = () => {
  const [products, setProducts] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);
  const endpoint = "http://localhost:8000/products";

  const fetchProducts = async () => {
    setIsLoading(true);
    setError(null);

    try {
      const response = await axios.get(endpoint);
      let updatedResponse = response.data.map((obj) => {
        delete obj["description"];
        delete obj["category"];
        return obj;
      });
      setProducts(updatedResponse);
    } catch (e) {
      setError(e);
    }
    setIsLoading(false);
  };

  useEffect(() => {
    fetchProducts();
  }, []);

  return (
    <AppContainer>
      <HeaderBar />

      <HomeContainer>
        <h1>Products</h1>
        {isLoading && (
          <LoadingContainer>
            <CircularProgress />
          </LoadingContainer>
        )}
        <div>{error && <p>Oops... please try again!</p>}</div>
        <ProductsPage products={products} />
      </HomeContainer>
    </AppContainer>
  );
};

export default Home;
