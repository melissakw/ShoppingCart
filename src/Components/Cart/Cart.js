import React, { useState } from "react";
import { useSelector } from "react-redux";
import HeaderBar from "../HeaderBar/HeaderBar";
import EmptyCart from "./EmptyCart";
import CartDetails from "./CartDetails";
import CircularProgress from "@material-ui/core/CircularProgress";
import { AppContainer } from "../App.elements";
import {
  Loading,
  CartRow,
  CartSummary,
  ShipTime,
  CheckoutButton
} from "./Cart.elements";

const Cart = () => {
  const productsInCart = useSelector((state) => state.product.cartItems);
  const totalCartPrice =
    Math.round(
      productsInCart.reduce((a, c) => a + c.price * c.quantity, 0) * 100
    ) / 100;
  const numOfItemsInCart = useSelector((state) =>
    state.product.cartItems.reduce((a, c) => a + +c.quantity, 0)
  );
  const [isLoadingChanges, setIsLoadingChanges] = useState(false);

  return (
    <AppContainer>
      <HeaderBar />
      <h1>Shopping Cart</h1>
      <div>{numOfItemsInCart === 0 && <EmptyCart />}</div>
      <div>
        {isLoadingChanges && (
          <Loading>
            <CircularProgress />
          </Loading>
        )}
      </div>

      <CartRow>
        <CartDetails
          productsInCart={productsInCart}
          setIsLoadingChanges={setIsLoadingChanges}
        />
        <CartSummary>
          <h2>Order Summary</h2>
          <p>Subtotal: ${totalCartPrice}</p>
          <p>Shipping: TBD</p>
          <p>Estimated Tax: $0.00</p>
          <p>Total: ${totalCartPrice}</p>
          <ShipTime>Estimated Delivery: Feb 16-24</ShipTime>
          {/*date for demo only. Actual date to be calculated*/}
          <CheckoutButton>Checkout</CheckoutButton>
        </CartSummary>
      </CartRow>
    </AppContainer>
  );
};

export default Cart;
