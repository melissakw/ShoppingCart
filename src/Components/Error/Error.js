import React from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { GoHomeButton, HighlightText, BodyText } from "./Error.elements";
import { AppContainer } from "../App.elements";

const Error = () => {
  return (
    <AppContainer>
      <HighlightText>Oops... </HighlightText>
      <BodyText>The page you were looking for was not found.</BodyText>

      <Link to="/" data-testid="go-home-link">
        <GoHomeButton >Go back</GoHomeButton>
      </Link>
    </AppContainer>
  );
};

export default Error;
