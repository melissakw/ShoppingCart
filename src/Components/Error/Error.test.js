import React from 'react';
import {render, screen, fireEvent, cleanup} from "@testing-library/react";
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import Cart from '../Cart/Cart';
import Error from './Error';
import Home from '../Home/Home';
import { Provider } from "react-redux";
import { ConfigureStore } from "../../Redux/ConfigureStore";
import { MemoryRouter, BrowserRouter, Switch, Route, Link } from 'react-router-dom';

test("renders Error Page correctly", () => {

	const tree = renderer.create(
		<BrowserRouter>
			<Error />
		</BrowserRouter>
		).toJSON();
	expect(tree).toMatchSnapshot();
});

test('navigates Home when "Go Home" is clicked', () => {

	const { persistor, store } = ConfigureStore();

	const {getByText} = render(
		<Provider store={store}>
			<MemoryRouter>
				<Switch>
					<Route exact path="/" component={Home} />
					<Route path="/cart" component={Cart} />
					<Route component={Error} />
				</Switch>

				<Error />
			</MemoryRouter>
		</Provider>
	)

const goHomeLink = screen.getByText(/go back/i);
goHomeLink.dispatchEvent(new MouseEvent("click", {bubbles: true}));

expect(screen.getByText(/products/i)).toBeInTheDocument();

});
