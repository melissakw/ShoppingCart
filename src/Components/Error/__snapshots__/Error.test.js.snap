// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Error Page correctly 1`] = `
<div
  className="sc-bdfBwQ olWrH"
>
  <h1
    className="sc-fFubgz dtguBp"
  >
    Oops... 
  </h1>
  <h3
    className="sc-bkzZxe linzoE"
  >
    The page you were looking for was not found.
  </h3>
  <a
    data-testid="go-home-link"
    href="/"
    onClick={[Function]}
  >
    <button
      className="sc-fodVxV gBMCGh"
    >
      Go back
    </button>
  </a>
</div>
`;
